src/Compilers/CSharp/Portable/Binder/Binder.cs

```diff
@@ -705,6 +705,21 @@ internal static ObsoleteDiagnosticKind ReportDiagnosticsIfObsoleteInternal(Diagn
             return kind;
         }
 
+        internal static void ReportDiagnosticsIfUnmanagedCallersOnly(DiagnosticBag diagnostics, MethodSymbol symbol, Location location)
+        {
+            var unmanagedCallersOnlyAttributeData = symbol.UnmanagedCallersOnlyAttributeData;
+            if (unmanagedCallersOnlyAttributeData != null)
+            {
+                // Either we haven't yet bound the attributes of this method, or there is an UnmanagedCallersOnly present.
```

Comment:
>we haven't yet bound the attributes of this method [](start = 26, length = 50)

Is this common? That is, are we likely to generate a fair number of lazy diagnostics as a result?

Reply:
A decent number, but it'll only happen in places where we haven't decoded the method attributes yet (which I think should only be when the method is invoked in an attribute itself). Regardless, I do not see a way around this without potentially incurring cycles.

---
In reply to: [482253826](https://github.com/dotnet/roslyn/pull/47311#discussion_r482253826) [](ancestors = 482253826)

Reply:
> it'll only happen in places where we haven't decoded the method attributes yet (which I think should only be when the method is invoked in an attribute itself)

Will we hit this potentially when binding a call between source methods, if the caller is bound before the attributes on the method being called? Looking at `SourceMethodSymbolWithAttributes.ObsoleteAttributeData`, that attribute is an "early decoded" attribute. Is that to minimize this particular issue?

---
In reply to: [482279736](https://github.com/dotnet/roslyn/pull/47311#discussion_r482279736) [](ancestors = 482279736,482253826)

Reply:
> Is that to minimize this particular issue?

Frankly, I'm not entirely sure. We could potentially do a 2-phase decode of unmanagedcallersonly, where early we just decode the presence of it, and later we do the full decode. Let me know if you'd like me to try this approach. I'm assuming we can't fully get rid of the lazy approach, however, or Obsolete would have been able to fully get rid of it as well, and it clearly has not.

---
In reply to: [482584779](https://github.com/dotnet/roslyn/pull/47311#discussion_r482584779) [](ancestors = 482584779,482279736,482253826)

Reply:
Moved to a two-phase decode: the first phase in early decode is responsible for actually populating the attribute data. It can't issue errors, however, so I added a step to PostDecode to check for UnmanagedCallersOnly error cases.

---
In reply to: [482612925](https://github.com/dotnet/roslyn/pull/47311#discussion_r482612925) [](ancestors = 482612925,482584779,482279736,482253826)